### PR TITLE
deprecate `with QuantumTape`

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
We are in the midst of a major overhaul for `QuantumTape`. This deprecation needs to happen in 0.29 so that in 0.30, we can make `QuantumTape` the same as `QuantumScript` (and remove `QuantumScript`).

**Description of the Change:**
Deprecate the `__enter__` dunder of `QuantumTape`, so that users who do `with QuantumTape` are advised to instead construct tapes directly, or use an AnnotatedQueue if necessary.

**Benefits:**
We will be on the last step of the QuantumTape deprecation once this is done.

**Possible Drawbacks:**
This is going to be a quick deprecation. Although QuantumTape should only be used by developers, I suspect that some of our users have grown accustomed to using `with tape`. This has to stop soon, but perhaps this is too soon.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -50,6 +50,11 @@
 
 <h3>Deprecations</h3>
 
+* `with qml.tape.QuantumTape()` is deprecated. Please initialize `QuantumTape` instances directly instead.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:53]: Given the size of this deprecation, we should use a code example as well.

Comment (timmysilv) [reply doc/releases/changelog-dev.md:53]: in the changelog _and_ the deprecation file (where I added one already)?

**Comment by josh146** [comment]
> Although `QuantumTape` should only be used by developers, I suspect that some of our users have grown accustomed to using `with tape`

Minor addendum, but `with tape:` became public/user facing as part of the transforms API, and is documented as user facing in both the arXiv paper [Quantum computing with differentiable quantum transforms](https://arxiv.org/abs/2202.13414) and the blog post [How to write quantum function transforms in PennyLane](https://pennylane.ai/blog/2021/08/how-to-write-quantum-function-transforms-in-pennylane/).

**Comment by timmysilv** [comment]
@josh146 do you mean to say you don't want us to do this? because removing `with tape` is a strongly required step in this process, and it won't be restored even once `QuantumScript` is un-renamed back to `QuantumTape`

**Comment by josh146** [comment]
@timmysilv yes, we need treat `with tape:` as part of the public API and plan this more carefully, alongside input from collaborators/researchers/external developers we know are using it (cc @trbromley)

**Comment by timmysilv** [comment]
I don't think we're going to do this. We'll just use `QuantumScript` internally, and keep `QuantumTape` available for external users (with it also including the `with tape` functionality). If it's somehow possible to make it clearer that we are avoiding `QuantumTape` internally, that would be cool.

**Pull request closed without merging**